harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,386 @@
+        if self._num_micro_batches_per_slot is None:
+            raise det.errors.InvalidExperimentException(
+                "Please call wrap_model_engine before accessing num_micro_batches_per_slot."
+            )
+        return self._num_micro_batches_per_slot
+
+    def set_profiler(self, *args: List[str], **kwargs: Any) -> None:
+        """
+        Sets a torch profiler instance on the trial context to be called in _pytorch_trial
+        when training.
+        """
+        self.profiler = torch.profiler.profile(
+            on_trace_ready=torch.profiler.tensorboard_trace_handler(str(get_base_path({}))),
+            *args,
+            **kwargs,
+        )
+
+    def _set_determined_profiler(self, prof: profiler.ProfilerAgent) -> None:
+        self._determined_profiler = prof
+
+    @contextlib.contextmanager
+    def _record_timing(self, metric_name: str, accumulate: bool = False) -> Iterator[None]:
+        if not self._determined_profiler:
+            yield
+            return
+        with self._determined_profiler.record_timing(metric_name, accumulate):
+            yield
+
+    def _init_device(self) -> None:
+        self.n_gpus = len(self.env.container_gpus)
+        if self.distributed.size > 1:
+            if self.n_gpus > 0:
+                # We launch a separate process per GPU with LOCAL_RANK set by DeepSpeed's launcher.
+                # Each process needs to bind to a unique GPU.
+                self.device = torch.device("cuda", int(cast(str, os.environ.get("LOCAL_RANK"))))
+                torch.cuda.set_device(self.device)
+            else:
+                self.device = torch.device("cpu")
+        elif self.n_gpus > 0:
+            self.device = torch.device("cuda", 0)
+        else:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def to_device(self, data: pytorch._Data) -> pytorch.TorchData:
+        """Map data to the device allocated by the Determined cluster.
+
+        Since we pass an iterable over the dataloader to `train_batch` and `evaluate_batch`
+        for DeepSpeedTrial, the user is responsible for moving data to GPU if needed.  This is
+        basically a helper function to make that easier.
+        """
+        with self._record_timing("to_device", accumulate=True):
+            return pytorch.to_device(data, self.device, self._to_device_warned_types)
+
+    def is_epoch_start(self) -> bool:
+        """
+        Returns true if the current batch is the first batch of the epoch.
+
+        .. warning::
+            Not accurate for variable size epochs.
+        """
+        if self._current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        if self._epoch_len is None:
+            raise det.errors.InternalException("Training DataLoader uninitialized.")
+        return self._current_batch_idx % self._epoch_len == 0
+
+    def is_epoch_end(self) -> bool:
```

Comment:
With model parallelism, every rank is still working on the same batch_idx at each step.  Although we will not necessarily build the dataloader on every rank, we do a gather on the chief rank and take the min epoch_len before broadcasting to all ranks.  This will make sure that ranks agree on epoch end and batch_idx.